### DISCUSSION TITLE
Multiple buys in a row?

### DISCUSSION THREAD

**Discussion by bioluminesceme** [kernc/backtesting.py#688] [Q&A]
I'm probably doing something stupid, but I'm testing basic things and I don't understand why I'm getting more than 1 buy in a row.. despite my if conditions.

I'm getting 2 or more buys in a row. Shouldn't it buy with all available cash, and then wait for the sell condition - because there is no money left to buy with? (and my if condition is not met in the first place)
If I print the broker cash amount it does show that it did not actually buy anything I think, but there seems to be something wrong with the math there too. the self._broker._cash gets printed _after_ the buy or sell is executed, so shouldn't broker cash reflect that buy or sell? the first two prints show the initialamount..

It's a "buy when StochRSI is below 30, sell when StochRSI is over 70". 
 

    import pandas as pd
    import talib
    from backtesting import Backtest, Strategy

    # download ticker data CSV from Yahoo, save in /data folder 
    ticker = "ENTERTICKERSYMBOL" 
    
    initialamount = 1000
    tickerdata = pd.read_csv(f"data/{ticker}.csv", index_col="Date", parse_dates=True)

    print(ticker)
    
    class StochRSI(Strategy):
        stochrsisettings = [ 12, 9, 6, 3]
    
        def init(self):
            super().init()
            self.stoch = self.I(talib.STOCHRSI, self.data.Close, self.stochrsisettings[0], self.stochrsisettings[1], self.stochrsisettings[2], self.stochrsisettings[3])
        
        def next(self):
            super().next()
            current_date = self.data.index[-1]
            current_close = self.data.Close[-1]
        
            if not self.position and self.stoch < 30:
                print(f"{current_date} - Buy when StochRSI is below 30 at ${current_close}")
                self.buy()
                print(self._broker._cash)
            
            if self.position and self.stoch > 70:
                print(f"{current_date} - Sell when StochRSI is over 70 at ${current_close}")
                self.position.close()   
                print(self._broker._cash)   
        

    bt = Backtest(tickerdata, StochRSI, cash=initialamount)
    stats = bt.run()

    print(stats)

    bt.plot()

print output example (download last year of TSLA data from Yahoo)

2021-09-14 00:00:00 - Buy when StochRSI is below 30 at $744.48999
1000
2021-09-27 00:00:00 - Sell when StochRSI is over 70 at $791.359985
1000
2021-11-04 00:00:00 - Buy when StochRSI is below 30 at $1229.910034
1042.200012
2021-11-05 00:00:00 - Buy when StochRSI is below 30 at $1222.089966
1042.200012 
2021-11-08 00:00:00 - Buy when StochRSI is below 30 at $1162.939941
1042.200012
2021-11-09 00:00:00 - Buy when StochRSI is below 30 at $1023.5
1042.200012
(?!?!)
2021-11-22 00:00:00 - Sell when StochRSI is over 70 at $1156.869995
1042.200012 (? it sold at 1156-1023 = 133 profit why is the broker cash amount still the same?)

**Comment by kernc** [comment]
Your assumptions are slightly misguided. When you place an order via `buy()` or close the position via [`Position.close()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Position.close), the trading is actually done on next candle Open. This might not explain multiple buys in a row (i.e. why market-order orders aren't being filled), but have you tried [`Backtest(..., trade_on_close=True)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest)?

**Reply by bioluminesceme** [reply]
Ah right, of course! It checks the Close so will trade the next day - that's the broker cash mystery solved. 

(Except it still prints the same output if I change it to 

    bt = Backtest(tickerdata, StochRSI, cash=initialamount, trade_on_close=True)
    
But as long as it's internally doing the right math it's fine.

Does this also affect the 2+ buys in a row thing? Where it registers an "I should buy" on day 1 but something changes on day 2 that makes it so it gets 'skipped' ?
